### PR TITLE
feat: basic pre-commit-hook

### LINKED ISSUE #1858: Pre-commit hook
Hi, I want to contribute a pre-commit hook to run Trivy while doing git commit with pre-commit installed.

This issue is to track the effort in case anyone would like to collaborate.

I plan to start this hopefully within a couple of weeks. I'm currently pressed for some time to work on it, but I wanted to get the issue open in case anyone would like to discuss anything about it. :)

### PR DESCRIPTION
**Pull request by mxab**
## Description
This PR adds a pre-commit-hooks file which enables users trivy to be used in the https://pre-commit.com/#golang ecosystem

pre-commit compiles the trivy binaries on the fly

in this basic setup I create a plain `trivy` hook which needs to be configured on the user side and an opinionated `trivyfs` which will run the `trivy fs --exit-code 1 .`

eg:
```yaml
repos:
-   repo: https://github.com/aquasecurity/trivy.git
    rev: v0.30.0
    hooks:
    -   id: trivyfs
```

A potential feature for trivy would be to allow something like `trivy fs --files file1 file2 file/3.txt`. This could speedup the pre-commit validation as it supports the only staged file to be checked.

## Related issues
- Close #1858

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [x] I've added usage information (if the PR introduces new options)
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: docs/getting-started/installation.md
Comment (DmitriyLewen) [inline docs/getting-started/installation.md:268]: `pre-commit ` is developer helper, like `golangci-lint`.
I am not sure that we need to add instruction to docs.

File: .pre-commit-hooks.yaml
Comment (DmitriyLewen) [inline .pre-commit-hooks.yaml:7]: `Trivy` repository has many test files with vulnerabilities.
Scan result will always return exit code 1 for these files.
I think we can only select 1 file.

File: docs/getting-started/installation.md
Comment (DmitriyLewen) [inline docs/getting-started/installation.md:258]: can we move this and another logic into Makefile?
- `pre-commit` doesn't have install script, but we can add check that `pre-commit` is installed, as for [tiny-go](https://github.com/aquasecurity/trivy/blob/ed322ac6664306c13ed673e0fa7dd6f76ebc9841/Makefile#L53-L59).
- we can add script to create `.pre-commit-config.yaml` file(if file doesn't exist) and install this config.

File: docs/getting-started/installation.md
Comment (DmitriyLewen) [inline docs/getting-started/installation.md:263]: I think we can use latest version of a release or revision of the current commit.
What do you think?

File: docs/getting-started/installation.md
Comment (mxab) [reply docs/getting-started/installation.md:268]: I thought as developer if I wanted to know on how to get started integrating trivy as a pre-commit hook I would look in the "getting started / installation" section

File: .pre-commit-hooks.yaml
Comment (DmitriyLewen) [reply .pre-commit-hooks.yaml:7]: Sorry for misleading you.
Yesterday I made a mistake with the logic of using `pre-commit`.
Thanks for help me understand.

File: docs/getting-started/installation.md
Comment (DmitriyLewen) [reply docs/getting-started/installation.md:268]: Now i understand. You're right. This is a good place for it.

File: .pre-commit-hooks.yaml
Comment (DmitriyLewen) [inline .pre-commit-hooks.yaml:4]: Do we need this hook? I think users will only use `trivyfs` hook. What do you think?
P.S. I tried to use this hook but can't set `PATH` to `args`. I got  ` FATAL   multiple targets cannot be specified`.

File: docs/getting-started/installation.md
Comment (DmitriyLewen) [inline docs/getting-started/installation.md:265]: what if we add information about the possibility of using [Trivy flags](https://aquasecurity.github.io/trivy/v0.29.2/docs/references/cli/fs/)?
for example:
```
      args:
        - --severity=CRITICAL
        - --skip-dirs=./test       
```

File: .pre-commit-hooks.yaml
Comment (mxab) [reply .pre-commit-hooks.yaml:4]: Yeah, I also misread something in the pre-commit docs, the reason why it works for the `trivyfs` hook but not the `trivy` one is that I configured `trivyfs` with [pass_filenames](https://pre-commit.com/#hooks-pass_filenames) `: false`

pre-commit per default will not check all the files but only the staged one, therefore I think you got the error `FATAL multiple targets cannot be specified` because pre-commit executed trivy like this:
`trivy fs . staged-file1.txt  staged-file2.txt`


I'll remove the "plain hook" for now as it makes only sense if you can execute a command allows the appending of multiple files to check

**Comment by DmitriyLewen** [comment]
I think 3 hooks is overkill.
I think only the `fs` subcommand will be used. It is also always necessary to use `--exit-code 1`.
We can only use one flag with `entry: trivy fs --exit-code 1 .`.
If the user needs to exclude some directories, he can use the `-skip-dirs` and `--skip-files` flags.

If many users need a different subcommand or selection of directories to scan, we can always add/change a hook.

**Comment by mxab** [comment]
Ok now there is only `trivyfs` which allows configuration and `trivyfsall` for a simple setup

File: docs/getting-started/installation.md
Comment (DmitriyLewen) [inline docs/getting-started/installation.md:259]:
```suggestion
You can use [trivy fs flags](../../docs/references/cli/fs/) to configure Trivy.
Insert the required flags in the `args` field.

```

Comment (DmitriyLewen) [reply docs/getting-started/installation.md:259]: I think it should be said that all flags work.

**Comment by mxab** [comment]
@DmitriyLewen I can remove the `trivyfsall` hook. The idea was just to have one simple hook with a basic config.
It is not really configurable as it has the `.` at the end and from how I understand trivy fs cli interface the path is the final argument therefore you cannot mixin any futher `args`

**Comment by DmitriyLewen** [comment]
@mxab 
I checked this config:
``` yaml
repos:
  - repo: https://github.com/mxab/trivy
    rev: eb31ef229cbd1ed76b7e2becc6bec0614240e389
    hooks:
    - id: trivyfsall
      args:
        - --severity=HIGH
```

It works correctly(report contains only `HIGH` vulnerabilities)

**Comment by mxab** [comment]
I toyed arround with pre-commit and I still really don't understand why this works. There must be something different when between the bash/zsh and python subprocess module

I went through the pre-commit code and added some logging right before shoves it into the `subprocess.Popen(cmd,**kwargs)`
The `cmd` array contains still `'trivy', 'fs', '--exit-code', '1', '.', '--skip-dirs=.git/'`
Maybe cobra does some magic there.

If you want to only have one hook I now would vote to drop the trivyfsall as this want allow to set the path part. if you add another `.` in the args in the trivyfsall it will fail

**Comment by DmitriyLewen** [comment]
ok, let's remove `trivyfsall`.
Also, when you will correct docs, you can check docs with `make mkdocs-serve`?
If I'm not mistaken, one more empty line is needed [here](https://github.com/aquasecurity/trivy/pull/2490/commits/eb31ef229cbd1ed76b7e2becc6bec0614240e389#diff-6828264893dbf2a68733af90bf5996b5b8afec681e5d6f216edca2ec9c8d1fa5R258-R259).

**Comment by mxab** [comment]
Ok, removed it and updated the docs

**Review by DmitriyLewen — APPROVED** [review]

**Comment by mxab** [comment]
@DmitriyLewen should I rebase my branch?

**Comment by DmitriyLewen** [comment]
@mxab , yes, please

**Comment by mxab** [comment]
@DmitriyLewen is there anything else you require? I think rebasing it makes only sense when you really going to merge it :)

**Comment by DmitriyLewen** [comment]
Hello @mxab . 

> is there anything else you require? 

No, thanks. 
We need to get approval from @knqyf263.

@knqyf263 I approved this PR. Could you take a look when you have time, please.

**Comment by tspearconquest** [comment]
Hey, thanks for making this. I apologize as I haven't had proper time to work on this after all. This looks good to me, however I'm not a maintainer so I can't merge it. I would be interested in seeing a trivy-docker hook to run trivy from a docker image, and that was what I had planned to contribute. If you can make it, I'll definitely use it, and if not, then I'll open up a new pull request when I've got time to work on it.

Thanks again!

**Comment by natefive** [comment]
@mxab can you rebase this? @knqyf263 is there any reason why this stalled thanks :)

**Comment by mxab** [comment]
@natefive done

**Comment by proinsias** [comment]
@mxab – thanks for this!

**Comment by mxab** [comment]
@tspearconquest great idea,
I updated the pre-commit hooks to run in two modes `trivy-go` and `trivy-docker`

@DmitriyLewen does this make sense to you

**Comment by tspearconquest** [comment]
This looks good. I would only suggest one change, to use the `--severity` flag so that we only fail for medium and higher vulnerabilities.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by mxab** [comment]
@tspearconquest you can add such flags via the `args` field when using it
```
    hooks:
    -   id: trivyfs-go # or trivy-docker
        args:
          - --severity
          - ...
```

**Comment by tspearconquest** [comment]
LGTM

**Comment by tspearconquest** [comment]
I have some ideas for how to enhance this later on when I have more time to come back to it.

It would be great, as one idea, to be able to have a script pre-process the file inputs from pre-commit.

I could foresee a use case of building an in-repo Dockerfile into an image and then running a scan on the image, and another use case to perform secrets or sbom scanning only, without doing the full fs scan. But it's definitely a topic for a separate PR. This satisfies my original plan and I want to take a moment to give a shout out to @mxab for his dedication towards getting this merged.

Hope knqyf263 will approve it soon.

**Comment by DmitriyLewen** [comment]
Hey @mxab 

I thinks it is good idea!
Thanks for your work.

**Comment by mxab** [comment]
Maybe someone could reach out to @knqyf263 , the PR is open for quite a while now :)

**Comment by matthewflannery** [comment]
+1 on this, please can we get this PR over the line @knqyf263 - Trivy is quickly becoming an outstanding tool and being able to integrate this as early as possible, as easily as possible into a developers workflow would be amazing. @simar7 @DmitriyLewen @afdesk @behara

**Comment by urbanadventurer** [comment]
+1 LGTM

**Comment by knqyf263** [comment]
Sorry for the late reply. Trivy has many features now and we are cautious about putting new integrations upstream because we need to maintain that for the long term. You might feel this change is small, but small things make a big difference.

We will probably merge this PR in the future, but if you need the pre-commit shortly, I'd recommend creating your repository and open a PR to add the repository link here.
https://aquasecurity.github.io/trivy/v0.34/ecosystem/tools/#community-tools

**Comment by mxab** [comment]
Not the answer I was looking for after half a year 🥲, but here we go:
https://github.com/aquasecurity/trivy/pull/3203

The downside of the external version is that I can only use the docker container based approach
In the first version it only uses the latest trivy image

As I understood you correctly you may merge this PR in the future, therefore I would leave it open

**Comment by jmreicha** [comment]
I would be curious to know what the thought is about the downsides of maintaining this integration. Not enough help? Time? Something else? It would be great to have an "official" pre-commit hook, it may even drive greater adoption.

**Comment by itaysk** [comment]
Trivy is very popular and many integrations are requested/suggested every day. The core team doesn't have the capacity nor the knowledge to properly maintain and support every single integration and target ecosystem. Moreover, there can be different legitimate implementations for same scenarios or different opinions around how to integrate Trivy, and we don't want to be the judges of that. We want to encourage the community (e.g author of this PR) to create and share their solutions with Trivy. The  only think Trivy is a project can do is to make all of those integrations discoverable in one place. That's why we created the "Ecosystem" section in the documentation which lists (with your help) all of Trivy's different plugins/integrations/use cases in one place (if you have any feedback about that, please feel free to share, probably not in this PR though).

I'm going to close this PR since the pre-commit hook is now [included in the Ecosystem](https://github.com/aquasecurity/trivy/pull/3203) list. Thanks so much to @mxab and other contributors.

**Pull request closed without merging**
